File: credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int luhn(long num);
void company(long num);

int main(void)
{
    // Prompt user for input
    long num = get_long("Number: ");

    // Run through the Luhn's algorithm
    int sum = luhn(num);

    // If pass Luhn's algorithm, then check if match any credit card company and print company name
    if (sum % 10 == 0)
    {
        company(num);
    }

    else
    {
        printf("sum is: %i\n", sum);
        printf("INVALID\n");
    }
}

int luhn(long num)
{
    int sum1 = 0;
    int sum2 = 0;

    // Loop through the odd-digit number backwards, multiply each by 2, and add the products
    for (int i = 0; num != 0; i++)
    {
        // Get each single digit
        int d = num % 10;

        // Multiply each odd digit by 2 and sum the products all together
        if (i % 2 != 0 && i != 0)
        {
            int double_d = d * 2;

            // Add up each digit of a product if it is a two-digit product
            if (double_d > 9)
            {
                sum1 = sum1 + (double_d % 10) + (double_d / 10);
            }

            else
            {
                sum1 = sum1 + double_d;
            }
        }

        // Add up the rest of digits altogether
        else
        {
            sum2 = sum2 + d;
        }

        num = num / 10;
    }

    return (sum1 + sum2);
}

void company(long num)
{
    int i = 0;
    int two_d;
    int one_d;

    while (num != 0)
    {
        int d = num % 10;
        if (num > 10 && num < 100)
        {
            two_d = num;
        }
        if (num > 0 && num < 10)
        {
            one_d = num;
        }
        num = num / 10;
        i += 1;
    }

    if (i == 15 && (two_d == 34 || two_d == 37))
    {
        printf("AMEX\n");
    }

    else if (i == 16 && (two_d == 51 || two_d == 52 || two_d == 53 || two_d == 54 || two_d == 55))
    {
        printf("MASTERCARD\n");
    }

    else if ((i == 13 || i == 16) && (one_d == 4))
    {
        printf("VISA\n");
    }

    else
    {
        printf("INVALID\n");
    }
}

File: sentimental-credit/credit.py
from cs50 import get_string
import re


def main():
    number = get_string("Number: ")

    # check length
    length = len(number)
    if length not in [13, 15, 16]:
        print("INVALID")

    else:
        # Luhn's algorithm
        if length % 2 == 1:
            # (for an odd number of digits)
            sum = 0
            idx = length - 1
            while idx > 0:
                product = int(number[idx - 1]) * 2
                product_d = str(product)
                for i in range(len(product_d)):
                    sum += int(product_d[i])
                idx -= 2
            i = 0
            while i < length:
                sum += int(number[i])
                i += 2
            last_d = str(sum)[-1]
            if last_d == "0":
                check_brand(number)
            else:
                print("INVALID")

        else:
            # (for an even number of digits)
            sum = 0
            idx = length - 1
            while idx >= 0:
                product = int(number[idx - 1]) * 2
                product_d = str(product)
                for i in range(len(product_d)):
                    sum += int(product_d[i])
                idx -= 2
            i = 1
            while i < length:
                sum += int(number[i])
                i += 2
            last_d = str(sum)[-1]
            if last_d == "0":
                check_brand(number)
            else:
                print("INVALID")

# check brand


def check_brand(number):
    length = len(number)

    # Check for American Express
    if (length == 15) and (re.search("^34", number) or re.search("^37", number)):
        print("AMEX")

    # Check for MasterCard
    elif (length == 16) and (re.search("^51", number) or re.search("^52", number)
                             or re.search("^53", number) or re.search("^54", number)
                             or re.search("^55", number)):
        print("MASTERCARD")

    # Check for Visa
    elif (length == 13 or 16) and (re.search("^4", number)):
        print("VISA")

    else:
        print("INVALID")


if __name__ == "__main__":
    main()
